# Survival_Analysis

Set up R studio on AWS EC2 (https://docs.elmcomputing.io/ami/x86/rstudio-steps.html)

Data - TCGA Pancreatic Cancer RNA seq data (https://github.com/rmoffitt/pdacR/blob/master/R/parse_TCGA_PAAD.R)

Basics of Survival Analysis in R (http://www.sthda.com/english/wiki/survival-analysis-basics)

R shiny web app (http://bioinformatics.sdstate.edu/idep/)

Shiny code to refer - (https://www.youtube.com/watch?v=cOfKmgxBCso)

Day 1
- Create basic Kaplan M curve using (http://www.sthda.com/english/wiki/survival-analysis-basics)
- Create a shiny app (https://www.youtube.com/watch?v=cOfKmgxBCso) display title, any kind of button by clicking that button we should be able to see Kaplan M plot
- Upload the code on this github repo.

Day 2
- Try to implement (https://github.com/jeffreythy/survival-analysis-shiny-app/tree/master) repo using our lung dat and using 2 categories (status, sex)


RNA-seq + CHip-seq integrated analysis
- CHIP-seq data for PAAD cancer cell lines (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6122955/)
- RNA-seq data for PAAD cancer in also available in the same paper
- Integrated analysis (https://static-content.springer.com/esm/art%3A10.1186%2Fs12864-020-07205-6/MediaObjects/12864_2020_7205_MOESM1_ESM.pdf)
